lcls-twincat-motion/Library/POUs/Motion/FB_MotionStage.TcPOU

```diff
@@ -10,35 +15,64 @@ VAR
 	bExecute: BOOL;
 	bFwdHit: BOOL;
 	bBwdHit: BOOL;
-	fbExecFallingEdge: F_TRIG;
+	ftExec: F_TRIG;
+	rtExec: R_TRIG;
+	rtUserExec: R_TRIG;
+	rtTarget: R_TRIG;
 	fbSetEnables: FB_SetEnables;
+    bPosGoal: BOOL;
+    bNegGoal: BOOL;
 END_VAR
 ]]></Declaration>
     <Implementation>
       <ST><![CDATA[// Start with an accurate status
 stMotionStage.Axis.ReadStatus();
 
+rtUserExec(CLK := stMotionStage.bExecute);
+bPosGoal := stMotionStage.stAxisStatus.fActPosition < stMotionStage.fPosition;
+bNegGoal := stMotionStage.stAxisStatus.fActPosition > stMotionStage.fPosition;
+
 // Handle auto-enable timing
 CASE stMotionStage.nEnableMode OF
-	ENUM_StageEnableMode.STAGE_ENABLE_ALWAYS:
+	ENUM_StageEnableMode.ALWAYS:
 		stMotionStage.bEnable := TRUE;
 		bExecute := stMotionStage.bExecute;
-	ENUM_StageEnableMode.STAGE_ENABLE_NEVER:
+	ENUM_StageEnableMode.NEVER:
 		bExecute := stMotionStage.bExecute;
-	ENUM_StageEnableMode.STAGE_ENABLE_MOTION:
-		IF stMotionStage.bExecute THEN
-			stMotionStage.bEnable := TRUE;
+	ENUM_StageEnableMode.DURING_MOTION:
+		IF rtUserExec.Q THEN
+            IF bPosGoal THEN
+                IF stMotionStage.bAllForwardEnable THEN
+                    stMotionStage.bEnable := TRUE;
+                ELSE
+                    stMotionStage.bError := TRUE;
+                    stMotionStage.nErrorId := 16#4358;
+				END_IF
+            ELSIF bNegGoal THEN
+                IF stMotionStage.bAllBackwardEnable THEN
+                    stMotionStage.bEnable := TRUE;
+                ELSE
+                    stMotionStage.bError := TRUE;
+                    stMotionStage.nErrorId := 16#4357;
+				END_IF
+			END_IF
 		END_IF
 		bExecute := stMotionStage.bExecute AND stMotionStage.bEnableDone;
 END_CASE
 
 // Automatically fill the correct nCmdData for homing
-IF stMotionStage.nCommand = ENUM_EpicsMotorCmd.EPICS_MOTOR_HOME THEN
+IF stMotionStage.nCommand = ENUM_EpicsMotorCmd.HOME THEN
 	stMotionStage.nCmdData := stMotionStage.nHomingMode;
 END_IF
 
+// Update all enable booleans
 fbSetEnables(stMotionStage:=stMotionStage);
 
+// No moves allowed in error states
```

Comment:
Why isn't this at the top?

Reply:
In case we accumulate an error between the top and `FB_DriveVirtual`